### ISSUE TITLE
Wasm threading/shared memory support

### ISSUE THREAD

**Issue by dschuff** [WebAssembly/binaryen#1085]
We need toolchain support for wasm's upcoming support of threads/atomics/shared linear memory. The proposal repo is located at https://github.com/WebAssembly/threads (the important files are https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md and https://github.com/WebAssembly/threads/blob/master/proposals/threads/Globals.md)

I've already started working on this (#1069, #1077, #1082, #1083), but probably should have created this tracking issue first.

This requires several types of support in Binaryen, which we can track the state of here:
- [x] IR for shared memories, and new instructions
  - [x] Shared memories (#1069)
  - [x] Atomic instructions (atomic loads/stores, atomic RMW, cmpxchg, wait/wake) (#1082, #1077, #1083)
  - [x] Import/export of mutable globals (~~#1644~~ #1785)
  - [x] Sign-extension operators (#1167)
- [x] Text and binary encoding for shared memories and new instructions (currently everything that there is IR support for)
- [x] Validation support for new IR (#1090, #1092)
- [x] Optimizer support for new IR (#1094)
- [x] asm2wasm support
- ~~[ ] s2wasm support~~ (or s2wasm's replacement)
- [x] Interpreter support for new IR

**Comment by dschuff** [comment]
Currently there are no complete implementations of wasm threads in any JS VMs that I know of, but work has started in V8 and probably others too.

My highest-priority goal is to support end-to-end compilation with emscripten ASAP to test with the VMs. But of course if we end up getting things working before the VMs are ready then we can still probably do some testing with the interpreter. And of course this also requires some work on the emscripten side. Fortunately there's already some pthreads support for asm.js there, so hopefully it won't be too huge.

**Comment by lars-t-hansen** [comment]
> Currently there are no complete implementations of wasm threads in any JS VMs that I know of, but work has started in V8 and probably others too.

Work is going on in SpiderMonkey but given current workloads and summer vacation schedules we're not expecting to be done until late Q3 (and that's not a commitment, just a well-educated guess).

**Comment by Lachrista** [comment]
@dschuff 
I use -s USE_PTHREADS=1 -s WASM=1 to compile a wasm.but ,When I use the wasm file in firefox and Google, it's error reporting.
"Uncaught TypeError: [object Uint32Array] is not an integer shared typed array.
    at Atomics.store (<anonymous>)
    at Object.initMainThreadBlock "
I don't understand why this is, because it can be used normally in asm. Js.
This problem is very upsetting to me. I don't know how to solve it.

**Comment by dschuff** [comment]
Wasm support for threads depends on the webassembly threads proposal (tracked in https://github.com/WebAssembly/threads) and is not yet shipped by default in browsers. [This page]( https://github.com/kripken/emscripten/wiki/Pthreads-with-WebAssembly) describes how to turn wasm thread support on in Chrome and Firefox. Although since then, both browsers have also turned SharedArrayBuffer off by default too, so you need to turn that on as well. For chrome you can change the command line to include `--enable-features='WebAssembly SharedArrayBuffer'`

**Comment by Lachrista** [comment]
I don't think that's the reason.
Because I have already opened the corresponding configuration in Google, but it will still be wrong.
chrome://flags/ ,I've already set up
(1)shared-array-buffer
(2)All about webassembly
now ,Has been an error
"Uncaught TypeError: [object Uint32Array] is not an integer shared typed array.
at Atomics.store ()
at Object.initMainThreadBlock "
I don't know how to solve this problem.
